### DISCUSSION TITLE
Waiting for NativeDatabase() Encryption Key Test Results

### DISCUSSION THREAD

**Discussion by shinmantis** [simolus3/drift#2806] [Q&A] [answered]
Hello, I am using the example for an encrypted database provided in the github repository:  [https://github.com/simolus3/drift/blob/develop/examples/encryption/lib/database.dart](url)

Everything appears to be working fine but, for the app that I'm building I would like to be able to wait for NativeDatabase() to indicate whether or not the password supplied by the user is correct (this is assuming that db.sqlite file already exists). <br><br>  In the example,  this is determined in lines 59 & 60 <br> ` // Test that the key is correct by selecting from a table'`<br> `db.execute('select count(*) from sqlite_master');`. <br><br>I can make an "awaitable" request to the db which allows me resolve the responseof NativeDatabase such as `Future<List<TodoItem>> get allTodoItems => select(todoItems).get();` <br>  but I'm wondering if there is better way to achieve this behavior than just invoking an awaitable method after instantiating the db just to wait for the result output of the key test.  Hope this makes sense.

**Comment by simolus3** [comment, marked as answer]
Sure, I think you can use the underlying sqlite3 package to make this independent from the drift opening procedure. To keep it asynchronous, move the check to a background isolate:

```dart
Future<bool> isValidKey(String key) async {
  final path = await getApplicationDocumentsDirectory();
  final file = p.join(path.path, 'app.db.enc');
  if (await File(file).exists()) {
    return await Isolate.run(() async {
      final db = sqlite3.open(file);
      final escapedKey = key.replaceAll("'", "''");

      db.execute("pragma key = '$escapedKey'");

      try {
        // Test that the key is correct by selecting from a table
        db.execute('select count(*) from sqlite_master');
        return true;
      } on SqliteException {
        return false;
      }
    });
  } else {
    return true;
  }
}
```

You would then use this method in your app before accessing the drift database.

**Reply by shinmantis** [reply]
Tested! This works beautifully 💯

**Comment by shinmantis** [comment]
Awesome! Thanks for the response @simolus3   I'll give it a try and see how it works
